src/main/java/com/google/devtools/build/lib/remote/GrpcRemoteExecutorKeepalived.java

```diff
@@ -0,0 +1,308 @@
+        // retrying when received a unauthenticated error, and propagate to refreshIfUnauthenticated
+        // which will then call retrier again. It will reset the retry time counter so we could
+        // retry more than --remote_retry times which is not expected.
+        response = retrier
+            .execute(() -> Utils.refreshIfUnauthenticated(this::execute, callCredentialsProvider),
+                backoff);
+
+        // If no response from Execute(), use WaitExecution() in a "loop" which is implicated inside
+        // the retry block.
+        if (response == null) {
+          response = retrier.execute(
+              () -> Utils.refreshIfUnauthenticated(this::waitExecution, callCredentialsProvider),
+              backoff);
+        }
+      }
+
+      return response;
+    }
+
+    @Nullable
+    ExecuteResponse execute() throws IOException {
+      Preconditions.checkState(lastOperation == null);
+
+      try {
+        Iterator<Operation> streamOperations = executionBlockingStubSupplier.get().execute(request);
+
+        // We don't want to reset backoff for Execute() since if there is an error:
+        //   1. If happened before we received a first response, we want to ensure the retry counter
+        //      is increased and call Execute() again.
+        //   2. Otherwise, we will fallback to WaitExecution() loop.
+        return handleStreamOperations(streamOperations, /* resetBackoff */ false);
+      } catch (StatusRuntimeException e) {
+        if (lastOperation != null) {
+          // By returning null, we are going to fallback to WaitExecution() loop.
+          return null;
+        }
+        throw new IOException(e);
+      }
+    }
+
+    @Nullable
+    ExecuteResponse waitExecution() throws IOException {
+      Preconditions.checkState(lastOperation != null);
+
+      WaitExecutionRequest request = WaitExecutionRequest.newBuilder()
+          .setName(lastOperation.getName())
+          .build();
+      try {
+        Iterator<Operation> streamOperations = executionBlockingStubSupplier.get()
+            .waitExecution(request);
+        // We want to reset backoff for WaitExecution() so we can "infinitely" wait for the
+        // execution to complete as long as they are making progress (by returning response).
+        return handleStreamOperations(streamOperations, /* resetBackoff */ true);
+      } catch (StatusRuntimeException e) {
+        if (e.getStatus().getCode() == Code.NOT_FOUND) {
+          // Operation was lost on the server. Retry Execute.
+          lastOperation = null;
+          return null;
+        }
+        throw new IOException(e);
+      }
+    }
+
+    /**
+     * Process a stream of operations from Execute() or WaitExecution().
+     *
+     * @param resetBackoff reset backoff if true once received a response successfully.
+     */
+    ExecuteResponse handleStreamOperations(Iterator<Operation> streamOperations,
```

Comment:
nit: not that this is very important, but this method is still called `handleStreamOperations` and not `streamOperations` and you said that you renamed this. How about calling the method `streamOperations` and the argument `operations` or `operationStream`?

Reply:
Sorry, I thought you were talking about renaming the parameter `operations` to `streamOperations`. How about rename to `handleOperationStream`? Since there is a corresponding method called `handleOperation` below for processing a single operation.

Reply:
Done.